pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1322 @@
+				<DelegatorState<T>>::insert(&acc, Delegator::new(collator.clone(), amount));
+				Self::deposit_event(Event::Delegation(acc, amount, collator, new_total));
+			}
+			Ok(().into())
+		}
+		/// Leave the set of delegators and, by implication, revoke all ongoing
+		/// delegations
+		#[pallet::weight(0)]
+		pub fn leave_delegators(origin: OriginFor<T>) -> DispatchResultWithPostInfo {
+			let acc = ensure_signed(origin)?;
+			let delegator = <DelegatorState<T>>::get(&acc).ok_or(Error::<T>::DelegatorDNE)?;
+			for bond in delegator.delegations.0 {
+				Self::delegator_leaves_collator(acc.clone(), bond.owner.clone())?;
+			}
+			<DelegatorState<T>>::remove(&acc);
+			Self::deposit_event(Event::DelegatorLeft(acc, delegator.total));
+			Ok(().into())
+		}
+		/// Revoke an existing delegation
+		#[pallet::weight(0)]
+		pub fn revoke_delegation(origin: OriginFor<T>, collator: T::AccountId) -> DispatchResultWithPostInfo {
+			Self::delegator_revokes_collator(ensure_signed(origin)?, collator)
+		}
+		/// Bond more for delegators with respect to a specific collator
+		/// candidate
+		#[pallet::weight(0)]
+		pub fn delegator_bond_more(
+			origin: OriginFor<T>,
+			candidate: T::AccountId,
+			more: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let delegator = ensure_signed(origin)?;
+			let mut delegations = <DelegatorState<T>>::get(&delegator).ok_or(Error::<T>::DelegatorDNE)?;
+			let mut collator = <CollatorState<T>>::get(&candidate).ok_or(Error::<T>::CandidateDNE)?;
+			let _ = delegations
+				.inc_delegation(candidate.clone(), more)
+				.ok_or(Error::<T>::DelegationDNE)?;
+			T::Currency::reserve(&delegator, more)?;
+			let before = collator.total;
+			collator.inc_delegator(delegator.clone(), more);
+			let after = collator.total;
+
+			if collator.is_active() {
+				Self::update_active(candidate.clone(), collator.total);
+			}
+
+			<CollatorState<T>>::insert(&candidate, collator);
+			<DelegatorState<T>>::insert(&delegator, delegations);
+			Self::deposit_event(Event::DelegationIncreased(delegator, candidate, before, after));
+			Ok(().into())
+		}
+		/// Bond less for delegators with respect to a specific delegator
+		/// candidate
+		#[pallet::weight(0)]
+		pub fn delegator_bond_less(
+			origin: OriginFor<T>,
+			candidate: T::AccountId,
+			less: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let delegator = ensure_signed(origin)?;
+			let mut delegations = <DelegatorState<T>>::get(&delegator).ok_or(Error::<T>::DelegatorDNE)?;
+			let mut collator = <CollatorState<T>>::get(&candidate).ok_or(Error::<T>::CandidateDNE)?;
+			let remaining = delegations
+				.dec_delegation(candidate.clone(), less)
+				.ok_or(Error::<T>::DelegationDNE)?
+				.ok_or(Error::<T>::Underflow)?;
+
+			ensure!(remaining >= T::MinDelegation::get(), Error::<T>::DelegationBelowMin);
+			ensure!(
+				delegations.total >= T::MinDelegatorStk::get(),
+				Error::<T>::NomBondBelowMin
+			);
+
+			T::Currency::unreserve(&delegator, less);
```

Comment:
```suggestion
			let err_amt = T::Currency::unreserve(&delegator, less);
			debug_assert!(err_amt.is_zero());
```

Reply:
You will be surprised when having such a pattern can catch errors which would not be found normally in tests.